gk/co.c

```diff
@@ -0,0 +1,165 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "co.h"
+
+static struct gk_co *
+get_next_co(struct gk_co *this_co)
+{
+	/*
+	 * It is unlikely because as long as there is more than
+	 * one working coroutine, there is at least 50% chance that
+	 * @this_co is not the last working coroutine.
+	 */
+	if (unlikely(this_co->cos_list.next == &this_co->work->working_cos)) {
+		/* @this_co is the last working co. */
+		return list_first_entry(&this_co->work->working_cos,
+			struct gk_co, cos_list);
+	}
+	return list_next_entry(this_co, cos_list);
+}
+
+static inline void
+yield_next(struct gk_co *this_co)
+{
+	aco_yield_to(get_next_co(this_co)->aco);
+}
+
+static inline void
+reset_leftover(struct gk_co_leftover *leftover)
+{
+	leftover->fe = NULL;
+	leftover->fib = NULL;
+}
+
+/*
+ * CAUTION: callers must be aware that when this function returns,
+ * @this_co->work might have changed!
+ */
+static struct gk_co_task *
+get_the_very_first_task(struct gk_co *this_co, struct gk_co ***p_this_co_bucket)
+{
+	struct gk_co_work *work = this_co->work;
+	struct gk_co_task *first_task = NULL;
+
+	do {
+		struct gk_co_task *task, *temp_task;
+
+		while (unlikely(list_empty(&work->tasks))) {
+			/* There is no work to do. */
+
+			struct gk_co *next_co = get_next_co(this_co);
+
+			/* Make @this_co idle. */
+			list_del(&this_co->cos_list);
+			list_add_tail(&this_co->cos_list, &work->idle_cos);
+
+			if (likely(this_co != next_co)) {
+				/*
+				 * @this_co is NOT the last working coroutine.
+				 * Yield to the next coroutine.
+				 */
+				aco_yield_to(next_co->aco);
+			} else {
+				/*
+				 * No more work and no more working coroutines.
+				 * Return to the main coroutine.
+				 */
+				aco_yield();
+			}
+		}
+
+		/* Find a task to assign to @this_co. */
+		list_for_each_entry_safe(task, temp_task, &work->tasks,
+				task_list) {
+			struct gk_co *task_owner_co;
+			uint16_t task_masked_hash;
+
+			list_del(&task->task_list);
+			task_masked_hash = task->task_hash & work->hash_mask;
+			task_owner_co = work->hash_to_assigned_co
+				[task_masked_hash];
+			if (task_owner_co == NULL) {
+				if (first_task == NULL) {
+					/* A task to call our own. */
+					first_task = task;
+					/*
+					 * Take over the following tasks that
+					 * share the same masked hash.
+					 */
+					*p_this_co_bucket =
+						&work->hash_to_assigned_co
+						[task_masked_hash];
+					**p_this_co_bucket = this_co;
```

Comment:
As we have found a task here, shouldn't we ``break`` directly?

Reply:
Indeed, after many iterations of debugging, I figured out the bug: missing the ``break`` here leads to memory leakage.

As ``list_for_each_entry_safe()`` is a loop, when it finds the ``first task``, instead of jumping out the loop directly, it goes to the ``next`` task entry, however, the next one doesn't get processed correctly, so that the packet belonging to that task doesn't get processed. That's why we saw the mismatch in the measurement output (i.e., ``#pkts != #req + #declined + #dropped + #distributed``). With more and more packets accumulated, Gatekeeper even cannot receive new packets.

After fixing this bug, it runs correctly. I will post the evaluation details once it's done.

Reply:
Thank you very much for catching this bug!

My fix is a little different; see the code for details and explanation.

For our evaluation, your fix and mine should have the same performance because `leftover` is not being reused as I explain at [this comment](https://github.com/AltraMayor/gatekeeper/pull/370#discussion_r344437305).